### PR TITLE
Feature/17403 check forms assign to window

### PR DESCRIPTION
**Pull request by pris54**

### CODE REVIEW

File: admin/services/check-form.service.js [hunk 1]
```diff
@@ -225,6 +225,92 @@ const checkFormService = {
...
+   * Get unassigned forms for selected check window.
+   * @param checkWindowAssignedForms
+   * @returns {Promise<*>}
+   */
+  getUnassignedFormsForCheckWindow: async (checkWindowAssignedForms) => {
+    let checkFormData
+    let checkFormList = []
+
+    if (!checkWindowAssignedForms) {
+      return false
```
Comment (jon-shipley) [inline admin/services/check-form.service.js]: Would it be better to return an empty array here?

File: admin/services/check-form.service.js [hunk 2]
```diff
@@ -225,6 +225,92 @@ const checkFormService = {
...
+   * Get assigned forms for selected check window.
+   * @param checkWindowAssignedForms
+   * @returns {Promise<*>}
+   */
+  getAssignedFormsForCheckWindow: async (checkWindowAssignedForms) => {
+    let checkFormData
+    let checkFormList = []
+
+    if (!checkWindowAssignedForms) {
+      return false
```
Comment (jon-shipley) [inline admin/services/check-form.service.js]: Again here, would `[]` be better for the return?

File: admin/spec/service/check-window.service.spec.js [hunk 3]
```diff
@@ -139,4 +140,34 @@ describe('check-window.service', () => {
       })
     })
   })
+
+  describe('#getCurrentCheckWindowsAndCountForms', () => {
+    let fetchCurrentCheckWindowsStub
+
+    beforeEach(() => {
+      fetchCurrentCheckWindowsStub = sandbox.stub(checkWindowDataService, 'fetchCurrentCheckWindows').resolves(checkWindowsMock)
+      service = setupService(function () { return Promise.resolve(checkWindowsAndCountFormsMock) })
```
Comment (jon-shipley) [inline admin/spec/service/check-window.service.spec.js]: Shouldn't this service be the service you are testing and not a mock?

File: admin/views/test-developer/assign-check-forms-to-windows.ejs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    <div class="column-two-thirds">
+        <% if (messages && messages.info && messages.info.length) { %>
+        <div class="info-message">
+            <%= messages.info %>
+        </div>
+        <% } %>
+
+        <% if (messages && messages.error && messages.error.length) { %>
+        <div class="error-summary" role="group" aria-labelledby="error-summary-heading-1">
+            <h2 class="heading-medium error-summary-heading" id="error-summary-heading-1" tabindex="0">
```
Comment (jon-shipley) [inline admin/views/test-developer/assign-check-forms-to-windows.ejs]: tabindex?
Don't see that one in the govuk elements guidelines: https://govuk-elements.herokuapp.com/typography/#typography-headings

There should be a tab index of -1 on the `<div>` but none on the `<h2>`

File: admin/views/test-developer/assign-forms.ejs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        <div class="column-two-thirds">
+            <% if (messages && messages.info && messages.info.length) { %>
+            <div class="info-message">
+                <a href="#anchor-<%= messages.formName %>"><%= messages.info %></a>
+            </div>
+            <% } %>
+
+            <% if (messages && messages.error && messages.error.length) { %>
+            <div class="error-summary" role="group" aria-labelledby="error-summary-heading-1">
+                <h2 class="heading-medium error-summary-heading" id="error-summary-heading-1" tabindex="0">
```
Comment (jon-shipley) [inline admin/views/test-developer/assign-forms.ejs]: tab index again - see previous comment

File: admin/views/test-developer/assign-forms.ejs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        </div>
+
+        <div class="column-one-third add-title-margin">
+            <aside role="complementary" class="support-column">
+                <h3 class="heading-medium">Related</h3>
+                <nav role="navigation">
+                    <ul class="list">
+                        <li>
+                            <a href="/PDFs/MTC_administration_guidance_June-2017-trial.pdf" target="_blank">Guidance</a>
```
Comment (jon-shipley) [inline admin/views/test-developer/assign-forms.ejs]: Needs rel="noopener noreferrer"  for target="_blank"
https://paper.dropbox.com/doc/Links-xiWNcJk5Wpb8Sj6G4Prtl

File: admin/routes/test-developer.js [hunk 7]
```diff
@@ -2,97 +2,22 @@
...
-  } catch (error) {
-    return next(error)
-  }
-
-  res.redirect('upload-and-view-forms')
-})
+router.get('/assign-form-to-window', isAuthenticated(rolesConfig.ROLE_TEST_DEVELOPER), (req, res, next) => checkFormController.assignCheckFormsToWindowsPage(req, res, next))
+router.get('/assign-form-to-window/:checkWindowId', isAuthenticated(rolesConfig.ROLE_TEST_DEVELOPER), (req, res, next) => checkFormController.assignCheckFormToWindowPage(req, res, next))
+router.post('/assign-form-to-window', isAuthenticated(rolesConfig.ROLE_TEST_DEVELOPER), (req, res, next) => checkFormController.saveAssignCheckFormsToWindow(req, res, next))
+router.get('/unassign-forms/:checkWindowId', isAuthenticated(rolesConfig.ROLE_TEST_DEVELOPER), (req, res, next) => checkFormController.unassignCheckFormsFromWindowPage(req, res, next))
```
[the remark this answers is not in the document]
Comment (pris54) [reply admin/routes/test-developer.js]: This is the path to unassign forms, it's a 'remove' link as a GET request, that's how other remove links have been resolved too. Do you want it as a POST request in this case?

File: admin/services/check-form.service.js (hunk 1 above)
Comment (pris54) [reply admin/services/check-form.service.js]: OK, but this requires modifying the corresponding view too.

File: admin/services/check-form.service.js (hunk 2 above)
Comment (pris54) [reply admin/services/check-form.service.js]: Same as above, views needs editing too.

File: admin/spec/service/check-window.service.spec.js (hunk 3 above)
Comment (pris54) [reply admin/spec/service/check-window.service.spec.js]: Yes, that's correct. Fix done.

File: admin/views/test-developer/assign-check-forms-to-windows.ejs (hunk 4 above)
Comment (pris54) [reply admin/views/test-developer/assign-check-forms-to-windows.ejs]: Modified as requested.

File: admin/views/test-developer/assign-forms.ejs (hunk 5 above)
Comment (pris54) [reply admin/views/test-developer/assign-forms.ejs]: Modified as requested.

File: admin/views/test-developer/assign-forms.ejs (hunk 6 above)
Comment (pris54) [reply admin/views/test-developer/assign-forms.ejs]: Added to all the links with target="_blank" of this feature.

File: admin/routes/test-developer.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jon-shipley) [reply admin/routes/test-developer.js]: @GuyHarwood - want to comment?

I would suggest that in this and every other case where the database state is changed via GET we should convert to POST requests. 

Let's see what @GuyHarwood suggests before committing to this.

**Comment by pris54** [comment]
@jon-shipley Thanks. I'll wait for @GuyHarwood to get back on this. We may need to refactor the modal and other cases were GET was used too.

File: admin/routes/test-developer.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/routes/test-developer.js]: @pris54 @jon-shipley just discussed with Paula - this will be modified

**Pull request merged**